AssertionError asTerm called on not-a-Term type Api when unpickling type Api from TASTy

Extracted the second crash from https://github.com/scala/scala3/issues/22986

## Compiler version
3.6.4, 3.8.3

## Minimized code

Note: the 0.1.0+2-f4cb325b-SNAPSHOT seems to be no longer available; I reproduced by rebuilding the artifact from commit at https://github.com/Katrix/DataPrism/commit/f4cb325b
```scala
//> using scala 3.6.4
//> using repositories ivy2local
//> using dep net.katsstuff::dataprism-skunk:0.1.0+2-f4cb325b-SNAPSHOT
// Publish dep from this repo: `cd local/DataPrism && sbt --client "common/publishLocal; cats/publishLocal; fs2/publishLocal; skunk/publishLocal"` (commit f4cb325b).

package repro

import java.time.OffsetDateTime

import scala.util.NotGiven

type Id[A] = A

trait MapRes[F[_], R] {
  type K[_[_]]

  def toK(r: R): K[F]
  def fromK(k: K[F]): R
}

object MapRes extends LowPriorityMapRes {
  type Aux[F[_], R, K0[_[_]]] = MapRes[F, R] { type K[G[_]] = K0[G] }

  given [F[_], K0[_[_]]]: MapRes.Aux[F, K0[F], K0] = new MapRes[F, K0[F]] {
    type K[F0[_]] = K0[F0]

    override def toK(r: K0[F]): K[F]   = r
    override def fromK(k: K[F]): K0[F] = k
  }

  given [F[_], T <: NonEmptyTuple](
      using ev: Tuple.IsMappedBy[F][T]
  ): MapRes.Aux[F, T, [F0[_]] =>> Tuple.Map[Tuple.InverseMap[T, F], F0]] = new MapRes[F, T] {
    override type K[F0[_]] = Tuple.Map[Tuple.InverseMap[T, F], F0]

    override def toK(r: T): Tuple.Map[Tuple.InverseMap[T, F], F]   = ???
    override def fromK(k: Tuple.Map[Tuple.InverseMap[T, F], F]): T = ???
  }

  given [F[_], A]: MapRes.Aux[F, F[A], [F0[_]] =>> F0[A]] = new MapRes[F, F[A]] {
    override type K[F0[_]] = F0[A]

    override def toK(r: F[A]): K[F]   = r
    override def fromK(k: K[F]): F[A] = k
  }

  given recurTuple1[F[_], V, MRK[_[_]]](
      using mr: MapRes.Aux[F, V, MRK]
  ): MapRes.Aux[F, Tuple1[V], [F0[_]] =>> Tuple1[MRK[F0]]] = new MapRes[F, Tuple1[V]]:
    override type K[F0[_]] = Tuple1[MRK[F0]]

    override def toK(r: Tuple1[V]): K[F]   = ???
    override def fromK(k: K[F]): Tuple1[V] = ???
  end recurTuple1
}

trait LowPriorityMapRes {

  given recurTuple[F[_], H, T <: Tuple, KH[_[_]], KT[_[_]] <: Tuple](
      using MapRes.Aux[F, H, KH],
      MapRes.Aux[F, T, KT]
  ): MapRes.Aux[F, H *: T, [F0[_]] =>> KH[F0] *: KT[F0]] = new MapRes[F, H *: T]:
    override type K[F0[_]] = KH[F0] *: KT[F0]

    override def toK(r: H *: T): K[F]   = ???
    override def fromK(k: K[F]): H *: T = ???
  end recurTuple
}

trait PlatformBase { self =>
  sealed trait SqlNull
  object SqlNull extends SqlNull
  type Nullable[A] = A | SqlNull

  type Nullability[A] <: NullabilityBase[A]
  trait NullabilityBase[A]:
    type N[_]

  type DbValue[A] <: DbValueBase[A]
  trait DbValueBase[A] {
    def ===(that: DbValue[A])(using n: Nullability[A]): DbValue[n.N[Boolean]]
  }

  type Api <: BaseApi
  val Api: Api
  trait BaseApi {
    type DbValue[A] = self.DbValue[A]
    type Nullable[A] = self.Nullable[A]
    type Nullability[A] = self.Nullability[A]
  }
}

trait Platform extends PlatformBase { self =>
  trait Nullability[A] extends NullabilityBase[A]
  object Nullability {
    type Aux[A, N0[_]] = Nullability[A] { type N[B] = N0[B] }

    given notNull[A](using NotGiven[SqlNull <:< A]): Nullability.Aux[A, Id] = new Nullability[A]:
      type N[B] = B

    given nullable[A >: SqlNull]: Nullability.Aux[A, Nullable] = new Nullability[A]:
      type N[B] = B | SqlNull
  }

  type DbValue[A] <: DbValueBase2[A]
  object DbValue {
    def falseV: DbValue[Boolean] = ???
  }

  trait DbValueBase2[A] extends DbValueBase[A]

  extension [A](a: DbValue[A | SqlNull])(using NotGiven[SqlNull <:< A]) {
    def map[B](f: DbValue[A] => DbValue[B]): DbValue[B | SqlNull] = ???
    def getOrElse(default: DbValue[A]): DbValue[A]                = ???
  }

  type Api <: BaseApi & PlatformApi
  trait PlatformApi {
    extension [A](a: DbValue[A | SqlNull])(using NotGiven[SqlNull <:< A]) {
      def map[B](f: DbValue[A] => DbValue[B]): DbValue[B | SqlNull] = self.map(a)(f)
      def getOrElse(default: DbValue[A]): DbValue[A] = self.getOrElse(a)(default)
    }
    inline def DbValue: self.DbValue.type = self.DbValue
  }
}
object Platform extends Platform {
  type Api = BaseApi & PlatformApi
  object Api extends BaseApi, PlatformApi
}
// Second crash variant per #22986: import from companion object, not Api; omit crash1 line.
import dataprism.skunk.platform.PostgresSkunkPlatform.{DbValue, Nullable}

sealed trait Table[A[_[_]]]

trait Query[A[_[_]]] {
  def join[B[_[_]]](that: Query[B])(
      on: (A[DbValue], B[DbValue]) => DbValue[Boolean]
  ): Query[Query.InnerJoin[A, B]] = ???

  def join[B[_[_]]](that: Table[B])(
      on: (A[DbValue], B[DbValue]) => DbValue[Boolean]
  ): Query[Query.InnerJoin[A, B]] = ???

  inline def map[R](f: A[DbValue] => R)(using res: MapRes[DbValue, R]): Query[res.K] = ???

  def mapK[B[_[_]]](f: A[DbValue] => B[DbValue]): Query[B] = ???
}
object Query {
  type InnerJoin[A[_[_]], B[_[_]]] = [F[_]] =>> (A[F], B[F])

  def from[A[_[_]]](table: Table[A]): Query[A] = ???
}

object Identifier {
  opaque type Identifier[A[_[_]]] = Int
}
export Identifier.Identifier

case class Topic[F[_]](id: F[Identifier[Topic]])
object Topic {
  val table: Table[Topic] = ???
}

case class Post[F[_]](
    topicId: F[Identifier[Topic]],
    creatorId: F[Nullable[Identifier[User]]],
    createdAt: F[OffsetDateTime]
)
object Post {
  val table: Table[Post] = ???
}

case class User[F[_]](id: F[Identifier[User]])
object User {
  val table: Table[User] = ???
}

def error = {
  val maxQuery: Query[[F[_]] =>> (F[Identifier[Topic]], F[Nullable[OffsetDateTime]])] = ???

  val query = Query.from(Post.table).join(maxQuery)((p, mp) => ???).map((p, mp) => (mp._1, p))

  // val crash1 = query.join(User.table) { case ((_, p2), u) => p2.creatorId.map(_ === u.id).getOrElse(DbValue.falseV) }
  val works  = query.join(User.table) { case (t, u) => t._2.creatorId.map(_ === u.id).getOrElse(DbValue.falseV) }
}
```

## Output

<details>

```scala
[error] Could not read definition type Api in dataprism-common_3-0.1.0%2B2-f4cb325b-SNAPSHOT.jar(dataprism/platform/sql/value/SqlDbValues.tasty). Caused by the following exception:
[error] java.lang.AssertionError: assertion failed: asTerm called on not-a-Term type Api
[error]     scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
[error]     dotty.tools.dotc.core.Symbols$Symbol.asTerm(Symbols.scala:186)
[error]     dotty.tools.dotc.ast.tpd$TreeOps$.ensureHasSym$extension(tpd.scala:1194)
[error]     dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readNewDef(TreeUnpickler.scala:1003)
[error]     dotty.tools.dotc.core.tasty.TreeUnpickler$TreeReader.readIndexedDef(TreeUnpickler.scala:867)
[error]     dotty.tools.dotc.core.tasty.TreeUnpickler$Completer.complete(TreeUnpickler.scala:166)
[error]     dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:175)
[error]     dotty.tools.dotc.core.Denotations$Denotation.completeInfo$1(Denotations.scala:190)
[error]     dotty.tools.dotc.core.Denotations$Denotation.info(Denotations.scala:192)
[error]     dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:393)
[error]     dotty.tools.dotc.core.SymDenotations$SymDenotation.flags(SymDenotations.scala:66)
[error]     dotty.tools.dotc.core.SymDenotations$SymDenotation.isOneOf(SymDenotations.scala:116)
[error]     dotty.tools.dotc.core.SymDenotations$SymDenotation.isEffectivelyErased(SymDenotations.scala:1033)
[error]     dotty.tools.dotc.transform.PruneErasedDefs.transformSym(PruneErasedDefs.scala:39)
[error]     dotty.tools.dotc.core.DenotTransformers$SymTransformer.transform(DenotTransformers.scala:72)
[error]     dotty.tools.dotc.core.DenotTransformers$SymTransformer.transform$(DenotTransformers.scala:67)
[error]     dotty.tools.dotc.transform.PruneErasedDefs.transform(PruneErasedDefs.scala:26)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:836)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:853)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Symbols$Symbol.recomputeDenot(Symbols.scala:124)
[error]     dotty.tools.dotc.core.Symbols$Symbol.computeDenot(Symbols.scala:118)
[error]     dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:109)
[error]     dotty.tools.dotc.core.Symbols$Symbol.name(Symbols.scala:277)
[error]     dotty.tools.dotc.core.Scopes$MutableScope.unlink(Scopes.scala:348)
[error]     dotty.tools.dotc.core.Scopes$Scope.drop$1(Scopes.scala:174)
[error]     dotty.tools.dotc.core.Scopes$Scope.filteredScope$$anonfun$1(Scopes.scala:181)
[error]     scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]     scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]     scala.collection.IterableOnceOps.foreach(IterableOnce.scala:619)
[error]     scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:617)
[error]     scala.collection.AbstractIterator.foreach(Iterator.scala:1303)
[error]     dotty.tools.dotc.core.Scopes$Scope.filteredScope(Scopes.scala:171)
[error]     dotty.tools.dotc.core.TypeErasure.dotty$tools$dotc$core$TypeErasure$$apply(TypeErasure.scala:739)
[error]     dotty.tools.dotc.core.TypeErasure.eraseInfo(TypeErasure.scala:825)
[error]     dotty.tools.dotc.core.TypeErasure$.transformInfo(TypeErasure.scala:277)
[error]     dotty.tools.dotc.transform.Erasure.transform(Erasure.scala:99)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.goForward$1(Denotations.scala:836)
[error]     dotty.tools.dotc.core.Denotations$SingleDenotation.current(Denotations.scala:882)
[error]     dotty.tools.dotc.core.Symbols$Symbol.recomputeDenot(Symbols.scala:124)
[error]     dotty.tools.dotc.core.Symbols$Symbol.computeDenot(Symbols.scala:118)
[error]     dotty.tools.dotc.core.Symbols$Symbol.denot(Symbols.scala:109)
[error]     dotty.tools.dotc.core.Symbols$.toDenot(Symbols.scala:544)
[error]     dotty.tools.dotc.transform.Erasure$Typer.recur$1(Erasure.scala:771)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedSelect(Erasure.scala:807)
[error]     dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3498)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3607)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedTypeApply(Erasure.scala:826)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3543)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:853)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3523)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:853)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3523)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:853)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3523)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1448)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3531)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.typer.Typer.typedReturn(Typer.scala:2299)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3539)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1448)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3531)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.$anonfun$16(Typer.scala:1542)
[error]     dotty.tools.dotc.typer.Applications.harmonic(Applications.scala:2651)
[error]     dotty.tools.dotc.typer.Applications.harmonic$(Applications.scala:465)
[error]     dotty.tools.dotc.typer.Typer.harmonic(Typer.scala:151)
[error]     dotty.tools.dotc.typer.Typer.typedIf(Typer.scala:1545)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedIf(Erasure.scala:899)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3532)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3739)
[error]     dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3758)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1085)
[error]     dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1441)
[error]     dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1445)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3531)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedLabeled(Typer.scala:2249)
[error]     dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3516)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3607)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.typer.Typer.$anonfun$66(Typer.scala:2921)
[error]     dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:256)
[error]     dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2921)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:972)
[error]     dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3505)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3607)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3712)
[error]     dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3758)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1085)
[error]     dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1441)
[error]     dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1445)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3531)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.transform.Erasure$Typer.$anonfun$7(Erasure.scala:866)
[error]     dotty.tools.dotc.core.Decorators$.zipWithConserve(Decorators.scala:160)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedApply(Erasure.scala:866)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3523)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:2857)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedValDef(Erasure.scala:923)
[error]     dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3502)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3607)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3712)
[error]     dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3758)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1085)
[error]     dotty.tools.dotc.typer.Typer.typedBlockStats(Typer.scala:1441)
[error]     dotty.tools.dotc.typer.Typer.typedBlock(Typer.scala:1445)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3531)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.typer.Typer.$anonfun$66(Typer.scala:2921)
[error]     dotty.tools.dotc.inlines.PrepareInlineable$.dropInlineIfError(PrepareInlineable.scala:256)
[error]     dotty.tools.dotc.typer.Typer.typedDefDef(Typer.scala:2921)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedDefDef(Erasure.scala:972)
[error]     dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3505)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3607)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3712)
[error]     dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3758)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1085)
[error]     dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:3187)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedClassDef(Erasure.scala:1061)
[error]     dotty.tools.dotc.typer.Typer.typedTypeOrClassDef$1(Typer.scala:3511)
[error]     dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:3515)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3607)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:3712)
[error]     dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:3758)
[error]     dotty.tools.dotc.transform.Erasure$Typer.typedStats(Erasure.scala:1085)
[error]     dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:3320)
[error]     dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:3557)
[error]     dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:3608)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3685)
[error]     dotty.tools.dotc.typer.Typer.typed(Typer.scala:3690)
[error]     dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:3801)
[error]     dotty.tools.dotc.transform.Erasure.run(Erasure.scala:146)
[error]     dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:383)
[error]     scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]     scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]     scala.collection.immutable.List.foreach(List.scala:334)
[error]     dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:376)
[error]     dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:343)
[error]     scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
[error]     scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
[error]     scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1323)
[error]     dotty.tools.dotc.Run.runPhases$1(Run.scala:336)
[error]     dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:383)
[error]     dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:395)
[error]     dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
[error]     dotty.tools.dotc.Run.compileUnits(Run.scala:395)
[error]     dotty.tools.dotc.Run.compileSources(Run.scala:282)
[error]     dotty.tools.dotc.Run.compile(Run.scala:267)
[error]     dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
[error]     dotty.tools.xsbt.CompilerBridgeDriver.run(CompilerBridgeDriver.java:141)
[error]     dotty.tools.xsbt.CompilerBridge.run(CompilerBridge.java:22)
[error]     sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:91)
[error]     sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.compileSources$1(BloopHighLevelCompiler.scala:148)
[error]     sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$9(BloopHighLevelCompiler.scala:181)
[error]     scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]     sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$1(BloopHighLevelCompiler.scala:73)
[error]     bloop.tracing.NoopTracer$.trace(BraveTracer.scala:53)
[error]     sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.timed$1(BloopHighLevelCompiler.scala:72)
[error]     sbt.internal.inc.bloop.internal.BloopHighLevelCompiler.$anonfun$compile$8(BloopHighLevelCompiler.scala:181)
[error]     scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[error]     monix.eval.internal.TaskRunLoop$.startFull(TaskRunLoop.scala:81)
[error]     monix.eval.internal.TaskRestartCallback.syncOnSuccess(TaskRestartCallback.scala:101)
[error]     monix.eval.internal.TaskRestartCallback.onSuccess(TaskRestartCallback.scala:74)
[error]     monix.eval.internal.TaskExecuteOn$AsyncRegister$$anon$1.run(TaskExecuteOn.scala:71)
[error]     java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
[error]     java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
[error]     java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
[error]     java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
[error]     java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
[error]     java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
[error]   val works  = query.join(User.table) { case (t, u) => t._2.creatorId.map(_ === u.id).getOrElse(DbValue.falseV) }
```

</details>

## Expectation
Should not crash, at the very least.

cc @Katrix @prolativ

--- a/compiler/src/dotty/tools/dotc/ast/tpd.scala
+++ b/compiler/src/dotty/tools/dotc/ast/tpd.scala
@@ -1212,7 +1212,7 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
     def ensureHasSym(sym: Symbol)(using Context): Unit =
       if sym.exists && sym != tree.symbol then
         typr.println(i"correcting definition symbol from ${tree.symbol.showLocated} to ${sym.showLocated}")
-        tree.overwriteType(NamedType(sym.owner.thisType, sym.asTerm.name, sym.denot))
+        tree.overwriteType(NamedType(sym.owner.thisType, sym.name, sym.denot))
 
     def etaExpandCFT(using Context): Tree =
       def expand(target: Tree, tp: Type)(using Context): Tree = tp match

--- a/compiler/src/dotty/tools/dotc/typer/Typer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Typer.scala
@@ -3262,6 +3262,7 @@ class Typer(@constructorOnly nestingLevel: Int = 0) extends Namer
           report.error(em"`$name` cannot have type parameters, because it ranges over capture sets", rhs.srcPos)
         case _ =>
     val res = assignType(cpy.TypeDef(tdef)(name, rhs1), sym)
+    res.ensureHasSym(sym)
     if Feature.ccEnabled && attachCap then
       res.putAttachment(CaptureVar, ())
     res

--- a/tests/pos/i25801/Lib_1.scala
+++ b/tests/pos/i25801/Lib_1.scala
@@ -1,0 +1,10 @@
+// https://github.com/scala/scala3/issues/25801
+package i25801
+
+trait A { self: B =>
+  type X
+}
+
+trait B extends A {
+  type X
+}

--- a/tests/pos/i25801/Use_2.scala
+++ b/tests/pos/i25801/Use_2.scala
@@ -1,0 +1,4 @@
+// https://github.com/scala/scala3/issues/25801
+package i25801
+
+def foo: B = ???

